### PR TITLE
Microsevice Integration of Port Manager and Data-Plane Manager

### PR DESCRIPTION
**Pull request by chenpiaoping**
This PR proposes the following changes:
* Slightly refactor DPM contract for code consistency
* Integrate port manager with DPM based on latest contract
* Remove Port Manager dao layer for now

### CODE REVIEW

File: web/src/main/java/com/futurewei/alcor/web/entity/dataplane/InternalPortEntity.java [hunk 1]
```diff
@@ -15,24 +15,27 @@
...
-  private List<NeighborInfo> neighborIps;
+    @JsonProperty("neighbor_info")
+    private List<NeighborInfo> neighborInfos;
 
-  @JsonProperty("binding_host_ip")
-  private String bindingHostIP;
+    @JsonProperty("binding_host_ip")
+    private String bindingHostIP;
 
-  private Set<InternalSubnetEntityNB> subnetEntities;
```
Comment (xieus) [inline web/src/main/java/com/futurewei/alcor/web/entity/dataplane/InternalPortEntity.java:36]: @chenpiaoping Check a few latest commits that were not included here: https://github.com/futurewei-cloud/alcor/pull/244/commits, particularly this one https://github.com/futurewei-cloud/alcor/pull/244/commits/cdbd96e349b92732bad75f656d69fadb19529695

Comment (chenpiaoping) [reply web/src/main/java/com/futurewei/alcor/web/entity/dataplane/InternalPortEntity.java:36]: Why does InternalPortEntity need the fields vpcEntities and subnetEntities？Don't they conflict with vpcEntities and subnetEntities in NetworkConfiguration?

Comment (xieus) [reply web/src/main/java/com/futurewei/alcor/web/entity/dataplane/InternalPortEntity.java:36]: Good question. Asked the same question when I saw them at the first place. Those two fields are internal fields which are not exposed to other services (port manager don't have to pass to DPM. DPM uses those two fields for internal bookkeeping purpose. @haboy52581

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java [hunk 2]
```diff
@@ -18,62 +18,199 @@
...
-        return cache.getAll();
+    public Map<String, PortEntity> findAllPortEntities() throws CacheException {
+        return portCache.getAll();
     }
 
-    @Override
-    public void addItem(PortEntity portEntity) throws CacheException {
-        cache.put(portEntity.getId(), portEntity);
+    public void createPortAndNeighbor(PortEntity portEntity, NeighborInfo neighborInfo) throws Exception {
+        try (Transaction tx = portCache.getTransaction().start()) {
```
Comment (xieus) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java:65]: Lock on portCache may not be sufficient. If multiple ports (in the same vpc) are created in separate calls, those calls will operate on different portCache row but the same neighborCache row.

Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java:65]: This lock can lock multiple cache, It should be taken out of the cache.

Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java:65]: This lock can lock multiple caches, It should be taken out of the cache.

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java [hunk 3]
```diff
@@ -146,21 +205,25 @@ public PortWebJson createPort(String projectId, PortWebJson portWebJson) throws
...
 
             //Wait for all async functions to finish
             List<Object> entities = executor.joinAll();
             entities.add(portEntity);
 
-            //Build GoalState and Send it to DPM
+            NeighborInfo neighborInfo = null;
             if (portEntity.getBindingHostId() != null) {
-                Goalstate.GoalState goalState = GoalStateUtil.buildGoalState(entities, Common.OperationType.CREATE);
+                Map<String, NodeInfo> nodeInfoMap = getNodeInfos(entities);
```
Comment (xieus) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java:216]: Try to use "this.getNodeInfos" if possible.

File: web/src/main/java/com/futurewei/alcor/web/entity/dataplane/NeighborInfo.java [hunk 4]
```diff
@@ -26,4 +25,50 @@
 
   @JsonProperty("port_id")
   private String portId;
+
+  @JsonProperty("port_mac")
+  private String portMac;
```
Comment (xieus) [inline web/src/main/java/com/futurewei/alcor/web/entity/dataplane/NeighborInfo.java:30]: Confirmed. This is needed. @chenpiaoping @haboy52581

**Comment by xieus** [comment]
@chenpiaoping Seems that you will need to address quite a bit of code conflicts.

**Comment by chenpiaoping** [comment]
> @chenpiaoping Seems that you will need to address quite a bit of code conflicts.

Yeah, I'm working on it.

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java (hunk 2 above)
Comment (xieus) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java:65]: oh could you pls point me to the right codes that explain a bit more about multi-cache locking? Just out of curiosity.

Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/repo/PortRepository.java:65]: https://github.com/futurewei-cloud/alcor/blob/master/lib/src/main/java/com/futurewei/alcor/common/db/ignite/IgniteTransaction.java

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java (hunk 3 above)
Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java:216]: ok

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/PortManagerApplication.java [hunk 5]
```diff
@@ -1,13 +1,10 @@
 package com.futurewei.alcor.portmanager;
 
-import org.mybatis.spring.annotation.MapperScan;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.transaction.annotation.EnableTransactionManagement;
 
 @SpringBootApplication
-@EnableTransactionManagement
```
Comment (xieus) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/PortManagerApplication.java:9]: Do we disable temporarily for some reason?

File: services/port_manager/src/main/resources/application.properties [hunk 6]
```diff
@@ -17,12 +17,4 @@ microservices.ip.service.url=http://localhost:9004/ips
...
-#MyBatis configuration
-spring.datasource.driverClassName=org.apache.ignite.IgniteJdbcThinDriver
-spring.datasource.url=jdbc:ignite:thin://127.0.0.1
-spring.datasource.schema=classpath:sql/port.sql
-mybatis.configuration.map-underscore-to-camel-case=true
-#spring.datasource.username=your_username
-#spring.datasource.password=your_password
-logging.level.com.dao=debug
\ No newline at end of file
+microservices.node.service.url=http://localhost:9007/nodes
```
Comment (xieus) [inline services/port_manager/src/main/resources/application.properties:20]: Need to add dpm service url as in commit https://github.com/futurewei-cloud/alcor/pull/244/commits/5fce67884499629dc754f192b6d02f86b77b5c4d

Let me add it for you to kick off the integration.

**Pull request merged**